### PR TITLE
RCE fix: replaced child_process execSync with spawnSync

### PR DESCRIPTION
**Pull request by d3v53c**
### 📊 Metadata *

The npm-git-publish module is vulnerable against RCE due to user-supplied inputs being formatted and executed without proper validation

### ⚙️ Description *

Mitigating the arbitrary code execution by changing the execution means rather than using more library code for sanitizing. 

### 💻 Technical Description *

Every call to "child_process.execSync" api is replaced by "child_process.spawn" apis. The `spawn`  method spawns a new process using the given command, with command-line arguments in args. If omitted, args defaults to an empty array. This mitigates the arbitrary code execution vulnerability.

### 🐛 Proof of Concept (PoC) *

Create the following PoC file:

`// poc.js`
`var git = require('npm-git-publish');`
`git.publish('.', 'http://gihub.com ;touch HACKED; #')`

Check there aren't files called HACKED
Execute the following commands in another terminal:

`npm i npm-git-publish # Install affected module`
`node poc.js #  Run the PoC`

Recheck the files: now HACKED has been created

### 🔥 Proof of Fix (PoF) *

Before: 

rce-before

After:

rce-after

### 👍 User Acceptance Testing (UAT)

test

### CODE REVIEW

**Review by Mik317 — CHANGES_REQUESTED** [review]
Hi @d3v53c :smile:
thanks for the fix :+1: 

The line seems ok to me, but could you please fix also the issue in the following lines?:

* https://github.com/418sec/npm-git-publish/blob/8332bbbe32acce0331f1bf9a54b47acfcba51803/lib/publish.ts#L132
* https://github.com/418sec/npm-git-publish/blob/8332bbbe32acce0331f1bf9a54b47acfcba51803/lib/publish.ts#L181
* https://github.com/418sec/npm-git-publish/blob/8332bbbe32acce0331f1bf9a54b47acfcba51803/lib/publish.ts#L188
* https://github.com/418sec/npm-git-publish/blob/8332bbbe32acce0331f1bf9a54b47acfcba51803/lib/publish.ts#L193
* https://github.com/418sec/npm-git-publish/blob/8332bbbe32acce0331f1bf9a54b47acfcba51803/lib/publish.ts#L206

Also, I would suggest using `execFile` instead of `spawn` (please do let me know if you chose that one for a specific reason :+1: ) which avoids totally the RCE :smile:

Cheers,
Mik

**Comment by d3v53c** [comment]
> Hi @d3v53c 😄
> thanks for the fix 👍
> 
> The line seems ok to me, but could you please fix also the issue in the following lines?:
> 
> * https://github.com/418sec/npm-git-publish/blob/8332bbbe32acce0331f1bf9a54b47acfcba51803/lib/publish.ts#L132
> * https://github.com/418sec/npm-git-publish/blob/8332bbbe32acce0331f1bf9a54b47acfcba51803/lib/publish.ts#L181
> * https://github.com/418sec/npm-git-publish/blob/8332bbbe32acce0331f1bf9a54b47acfcba51803/lib/publish.ts#L188
> * https://github.com/418sec/npm-git-publish/blob/8332bbbe32acce0331f1bf9a54b47acfcba51803/lib/publish.ts#L193
> * https://github.com/418sec/npm-git-publish/blob/8332bbbe32acce0331f1bf9a54b47acfcba51803/lib/publish.ts#L206
> 
> Also, I would suggest using `execFile` instead of `spawn` (please do let me know if you chose that one for a specific reason 👍 ) which avoids totally the RCE 😄
> 
> Cheers,
> Mik

Hi @Mik317, Good morning,

I chose `spawn` over `execFile`, because `execFile` runs the executable until it exits or terminates, then returns a buffer for data on stdout or stderr with a maximum size of 200Kb. `spawn` can stream stdout or stderr back to the parent process once it starts running, and there is no limit to the size of data it can return. Also, I'm a little biased towards `spawn`. 😄

These 2 functions apparently run in a largely similar manner, by executing a file in a child process. I didn't have any specific reason other than this. If you insist, I can change it to `execFile`, since in this case, it hardly matters, I guess.

Please let me know your thoughts too.

Cheers!!

**Comment by Mik317** [comment]
> > Hi @d3v53c smile
> > thanks for the fix +1
> > The line seems ok to me, but could you please fix also the issue in the following lines?:
> > 
> > * https://github.com/418sec/npm-git-publish/blob/8332bbbe32acce0331f1bf9a54b47acfcba51803/lib/publish.ts#L132
> > * https://github.com/418sec/npm-git-publish/blob/8332bbbe32acce0331f1bf9a54b47acfcba51803/lib/publish.ts#L181
> > * https://github.com/418sec/npm-git-publish/blob/8332bbbe32acce0331f1bf9a54b47acfcba51803/lib/publish.ts#L188
> > * https://github.com/418sec/npm-git-publish/blob/8332bbbe32acce0331f1bf9a54b47acfcba51803/lib/publish.ts#L193
> > * https://github.com/418sec/npm-git-publish/blob/8332bbbe32acce0331f1bf9a54b47acfcba51803/lib/publish.ts#L206
> > 
> > Also, I would suggest using `execFile` instead of `spawn` (please do let me know if you chose that one for a specific reason +1 ) which avoids totally the RCE smile
> > Cheers,
> > Mik
> 
> Hi @Mik317, Good morning,
> 
> I chose `spawn` over `execFile`, because `execFile` runs the executable until it exits or terminates, then returns a buffer for data on stdout or stderr with a maximum size of 200Kb. `spawn` can stream stdout or stderr back to the parent process once it starts running, and there is no limit to the size of data it can return. Also, I'm a little biased towards `spawn`. 
> 
> These 2 functions apparently run in a largely similar manner, by executing a file in a child process. I didn't have any specific reason other than this. If you insist, I can change it to `execFile`, since in this case, it hardly matters, I guess.
> 
> Please let me know your thoughts too.
> 
> Cheers!!

Thanks for the explanation :smile:
I think `spawn` is ok :+1: 
As I previously said, if you can fix also the other lines which could present the vulnerability it would be awesome :100: (if you re-ask the review I can `accept` those modifications too)

Cheers,
Mik

**Comment by d3v53c** [comment]
> > > Hi @d3v53c smile
> > > thanks for the fix +1
> > > The line seems ok to me, but could you please fix also the issue in the following lines?:
> > > 
> > > * https://github.com/418sec/npm-git-publish/blob/8332bbbe32acce0331f1bf9a54b47acfcba51803/lib/publish.ts#L132
> > > * https://github.com/418sec/npm-git-publish/blob/8332bbbe32acce0331f1bf9a54b47acfcba51803/lib/publish.ts#L181
> > > * https://github.com/418sec/npm-git-publish/blob/8332bbbe32acce0331f1bf9a54b47acfcba51803/lib/publish.ts#L188
> > > * https://github.com/418sec/npm-git-publish/blob/8332bbbe32acce0331f1bf9a54b47acfcba51803/lib/publish.ts#L193
> > > * https://github.com/418sec/npm-git-publish/blob/8332bbbe32acce0331f1bf9a54b47acfcba51803/lib/publish.ts#L206
> > > 
> > > Also, I would suggest using `execFile` instead of `spawn` (please do let me know if you chose that one for a specific reason +1 ) which avoids totally the RCE smile
> > > Cheers,
> > > Mik
> > 
> > 
> > Hi @Mik317, Good morning,
> > I chose `spawn` over `execFile`, because `execFile` runs the executable until it exits or terminates, then returns a buffer for data on stdout or stderr with a maximum size of 200Kb. `spawn` can stream stdout or stderr back to the parent process once it starts running, and there is no limit to the size of data it can return. Also, I'm a little biased towards `spawn`.
> > These 2 functions apparently run in a largely similar manner, by executing a file in a child process. I didn't have any specific reason other than this. If you insist, I can change it to `execFile`, since in this case, it hardly matters, I guess.
> > Please let me know your thoughts too.
> > Cheers!!
> 
> Thanks for the explanation 😄
> I think `spawn` is ok 👍
> As I previously said, if you can fix also the other lines which could present the vulnerability it would be awesome 💯 (if you re-ask the review I can `accept` those modifications too)
> 
> Cheers,
> Mik

Thank you @Mik317,
I'll update the PR once I have done with the requested changes, and I'll re-ask for your review too.

You're awesome!! 
Advanced happy new year wishes to you and all..!!

Cheers!!

**Comment by Mik317** [comment]
Thanks :heart:
Happy new year you too :+1: 

Cheers,
Mik

File: lib/publish.ts [hunk 1]
```diff
@@ -129,10 +152,12 @@ function doPublish(packageDir: string, gitRemoteUrl: string, params: Params): Pr
 
     function packPackageIntoTarball() {
         return directoryReady
-            .then(() => exec(`npm pack "${packageDir}"`, { cwd: packDir }))
+            .then(() => run([`npm`, `pack`, `"${packageDir}"`], { cwd: packDir }))
```
Comment (Mik317) [inline lib/publish.ts]: The `"` make the command not work correctly

File: lib/publish.ts [hunk 2]
```diff
@@ -178,20 +203,22 @@ function doPublish(packageDir: string, gitRemoteUrl: string, params: Params): Pr
...
-                return exec(`git tag -a --file="${tagTextPath}" "${tagName}"`, { cwd: gitRepoDir })
+                return parseOut(run([`git`, `tag`, `-a`, `--file="${tagTextPath}"`, `"${tagName}"`], { cwd: gitRepoDir }))
                 .then(() => {
                     let promises: Promise<string>[] = [];
                     (params.extraBranchNames || []).forEach((extraBranchName) => {
-                        promises.push(exec(
-                            `git branch -f "${extraBranchName}" "${tagName}"`,
+                        promises.push(parseOut(
+                          run(
+                            [`git`, `branch`, `-f`, `"${extraBranchName}"`, `"${tagName}"`],
```
Comment (Mik317) [inline lib/publish.ts]: `"` make errors

**Review by Mik317 — CHANGES_REQUESTED** [review]
Hi :smile:
I check and it seems really good but unfortunately the `double quotes` don't make the command work properly:

Screenshot from 2021-01-07 13-56-47

`"` seems to be useless since every `string` in array is considered to be `argument` :+1: 

Cheers,
Mik

File: lib/publish.ts (hunk 1 above)
Comment (d3v53c) [reply lib/publish.ts]: I have removed the quotes and reworked a little bit for synchronous execution.

File: lib/publish.ts (hunk 2 above)
Comment (d3v53c) [reply lib/publish.ts]: I have removed the quotes and reworked a little bit for synchronous execution.

File: lib/publish.ts [hunk 3]
```diff
@@ -129,10 +142,12 @@ function doPublish(packageDir: string, gitRemoteUrl: string, params: Params): Pr
 
     function packPackageIntoTarball() {
         return directoryReady
-            .then(() => exec(`npm pack "${packageDir}"`, { cwd: packDir }))
+            .then(() => run([`npm`, `pack`, `${packageDir}`], { cwd: packDir }))
             .then(() => {
                 // pack succeeded! Schedule a cleanup and return the full path
-                cleanupOperations.push(exec(`npm cache clean ${params.originalPackageInfo.name}@${params.originalPackageInfo.version}`));
+                cleanupOperations.push(
+                  parseOut(run([`npm`, `cache`, `clean`, `${params.originalPackageInfo.name}@${params.originalPackageInfo.version}`]))
```
Comment (d3v53c) [inline lib/publish.ts:149]: @Mik317 , this line raises an exception saying `npm cache clear` doesn't accept any arguments. If the goal here is to verify that the cache is consistent, then we could use `npm cache verify` as it's the one recommended by `npm` itself. I'd love to hear your thoughts on this, or I guess we could leave it to the maintainers hands, I suppose.

**Comment by JamieSlome** [comment]
@d3v53c - thanks for your PR.

This bounty was made live during our learning experiment on the platform where we encouraged our users to submit fixes for previously fixed vulnerabilities.

This vulnerability was previously fixed [here](https://github.com/418sec/npm-git-publish/pull/1)...

Apologies for the delay in the response! 🍰

**Comment by huntr-helper** [comment]
Sorry d3v53c, we enjoyed reviewing your fix but it has not been selected this time. If this bounty has not been closed, please feel free to try again with a new pull request!

We appreciate your effort and look forward to reviewing more of your fixes in the future! 🔨😎

[Come join us on Discord](https://huntr.dev/discord)

**Pull request closed without merging**